### PR TITLE
Fukui and Dyson analysis

### PR DESCRIPTION
**Pull request by tmhuysen**
**Short description**
Adds a black box module for Dyson and Fukui analysis with a pybinding.

**Related issues**
#422

**Don't forget:**

- [x]  if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

### CODE REVIEW

**Comment by tmhuysen** [comment]
How does one update the Doxygen file?

**Comment by lelemmen** [comment]
> How does one update the Doxygen file?

@tmhuysen 'updating' the Doxygen file is just making sure that new folders are registered in `Doxygen.in` from line 794 and onwards.

File: include/QCMethod/FukuiDysonAnalysis.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#include "Operator/SecondQuantized/SQHamiltonian.hpp"
+#include "RDM/RDMCalculator.hpp"
+
+
+namespace GQCP {
+namespace QCMethod {
+
+
+/**
+ *  A class that solves the FCI Hamiltonian for a given molecule and performs Fukui and Dyson analysis
```
Comment (lelemmen) [inline include/QCMethod/FukuiDysonAnalysis.hpp]:
```suggestion
 *  A class that diagonalizes the FCI Hamiltonian for a given molecule and performs Fukui and Dyson analysis
```

explain in detail what is meant by 'Fukui and Dyson analysis'

File: include/QCMethod/FukuiDysonAnalysis.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ *  restricted Hartree-Fock is initially performed for the singlet,
```
Comment (lelemmen) [inline include/QCMethod/FukuiDysonAnalysis.hpp]:
```suggestion
 * The same orthonormal basis is used for both species (N and N-1 number of electrons): RHF is used to determine an orthonormal basis for the species with an even number of electrons
```

File: include/QCMethod/FukuiDysonAnalysis.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+
+namespace GQCP {
+namespace QCMethod {
+
+
+/**
+ *  A class that solves the FCI Hamiltonian for a given molecule and performs Fukui and Dyson analysis
+ *  restricted Hartree-Fock is initially performed for the singlet, 
+ *  after which we solve in the FCI space and solve in the FCI space again after removing an electron
```
Comment (lelemmen) [inline include/QCMethod/FukuiDysonAnalysis.hpp]:
```suggestion
```

File: include/QCMethod/FukuiDysonAnalysis.hpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ */
+class FukuiDysonAnalysis {
+private:
+    Molecule molecule;
+    SingleParticleBasis<double, GTOShell> sp_basis;
+    SQHamiltonian<double> sq_hamiltonian;
+    std::string basis_set;  // the basisset that should be used
+    ProductFockSpace fock_space1 = ProductFockSpace(0, 0, 0); // Default
```
Comment (lelemmen) [inline include/QCMethod/FukuiDysonAnalysis.hpp]:
```suggestion
    ProductFockSpace fock_space1 = ProductFockSpace(0, 0, 0);  // default
```

File: include/QCMethod/FukuiDysonAnalysis.hpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    ProductFockSpace fock_space2 = ProductFockSpace(0, 0, 0); // Default
```
Comment (lelemmen) [inline include/QCMethod/FukuiDysonAnalysis.hpp]: idem

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        sp_basis (SingleParticleBasis<double, GTOShell>(molecule, basis_set)),
+        sq_hamiltonian (SQHamiltonian<double>::Molecular(this->sp_basis, molecule)),  // in AO basis
+        basis_set (basis_set)
+{
+    const auto K = this->sp_basis.numberOfBasisFunctions();
+    const auto N_P = this->molecule.numberOfElectrons()/2;
+
+    // Define a molecule on which an RHF calculation is allowed
+    auto restricted_molecule = this->molecule; 
+    const bool open_shell_entry = (N_P*2 != this->molecule.numberOfElectrons());  // If the entry is open shell, we require to create a closed shell molecule to allow RHF
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]:
```suggestion
    const bool open_shell_entry = (N_P*2 != this->molecule.numberOfElectrons());  // if the entry has an odd number of electrons, we create a molecule with an even number of electrons to allow RHF
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        auto rhf_solution = diis_scf_solver.get_solution();
+        basisTransform(this->sp_basis, this->sq_hamiltonian, rhf_solution.get_C());
+    } else {
+        PlainRHFSCFSolver plain_scf_solver (this->sq_hamiltonian, this->sp_basis, restricted_molecule);
+        plain_scf_solver.solve();
+        auto rhf_solution = plain_scf_solver.get_solution();
+        basisTransform(this->sp_basis, this->sq_hamiltonian, rhf_solution.get_C());
+    }
+
+    // Depending on wether the initial molecule was open shell we chose different Fock spaces as fock_space1 should always have the highest occupation to fit the algorithm
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]: rephrase
```suggestion
    // In order to supply the correct arguments to the algorithm ... 
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    if (open_shell_entry) {
+        this->fock_space1 = ProductFockSpace(K, N_P + 1, N_P);
+        this->fock_space2 = ProductFockSpace(K, N_P, N_P);
+    } else {
+        this->fock_space1 = ProductFockSpace(K, N_P, N_P);
+        this->fock_space2 = ProductFockSpace(K, N_P, N_P - 1); 
+    }
+
+    // Solve in the FCI space
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]:
```suggestion
    // Solve the FCI Hamiltonian eigenvalue problems with a Davidson algorithm
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    FCI fci1 = FCI(fock_space1);
+    FCI fci2 = FCI(fock_space2);
+    
+    CISolver ci_solver1(fci1, sq_hamiltonian);
+    CISolver ci_solver2(fci2, sq_hamiltonian);
+
+    // Solve with Davidson
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]:
```suggestion
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    // Retrieve the wavefunctions
+    const auto wavefunction1 = ci_solver1.makeWavefunction();
+    const auto wavefunction2 = ci_solver2.makeWavefunction();
+
+    // Extract dyson coefficients
+    this->dyson_coefficients = calculateDysonOrbitalCoefficients(wavefunction1, wavefunction2);
+    
+    // Set up calculations for Fukui analysis by calculating the 1RDMs
+    RDMCalculator rdm_calculator1 (fock_space1);
+
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]:
```suggestion
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    rdm_calculator1.set_coefficients(wavefunction1.get_coefficients());
+
+    RDMCalculator rdm_calculator2 (fock_space2);
+
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]:
```suggestion
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    ci_solver2.solve(davidson_solver_options2);
+
+    // Retrieve the wavefunctions
+    const auto wavefunction1 = ci_solver1.makeWavefunction();
+    const auto wavefunction2 = ci_solver2.makeWavefunction();
+
+    // Extract dyson coefficients
+    this->dyson_coefficients = calculateDysonOrbitalCoefficients(wavefunction1, wavefunction2);
+    
+    // Set up calculations for Fukui analysis by calculating the 1RDMs
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]:
```suggestion
    // Calculate the Fukui matrix
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    rdm_calculator1.set_coefficients(wavefunction1.get_coefficients());
+
+    RDMCalculator rdm_calculator2 (fock_space2);
+
+    rdm_calculator2.set_coefficients(wavefunction2.get_coefficients());
+
+    const auto onerdm1 = rdm_calculator1.calculate1RDMs().one_rdm;
+    const auto onerdm2 = rdm_calculator2.calculate1RDMs().one_rdm;
+
+    // The Fukui 1RDM
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]:
```suggestion
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    this->fukui_matrix = OneRDM<double>(onerdm1 - onerdm2);
+
+    // Diagonalize the fukui_matrix to retrieve Fukui naturals
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]:
```suggestion
    // Diagonalize the Fukui matrix to retrieve Fukui naturals
```

File: src/QCMethod/FukuiDysonAnalysis.cpp [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+    RDMCalculator rdm_calculator2 (fock_space2);
+
+    rdm_calculator2.set_coefficients(wavefunction2.get_coefficients());
+
+    const auto onerdm1 = rdm_calculator1.calculate1RDMs().one_rdm;
+    const auto onerdm2 = rdm_calculator2.calculate1RDMs().one_rdm;
+
+    // The Fukui 1RDM
+    this->fukui_matrix = OneRDM<double>(onerdm1 - onerdm2);
```
Comment (lelemmen) [inline src/QCMethod/FukuiDysonAnalysis.cpp]: I didn't know this cast was required, nice!

Comment (tmhuysen) [reply src/QCMethod/FukuiDysonAnalysis.cpp]: I don't think it is required

**Pull request merged**